ENCD-6191-listed-carts

forresttanaka opened the pull request:
* Calling the cart menu “status” got pretty confusing when cart status as a property of a cart has become so important, so I moved much of status.js into status_listing.js, and use status.js specifically for things relating directly to cart status.
* I also moved a lot of functionality out of index.js into store.js, as index.js’s job is mostly as an interface to other modules; not to implement functionality.
* Much of the other changes to seemly unrelated cart components have to do with released carts which can’t be altered by anyone, including their owners.

https://encd-6191-cart-sharing-pv1-fytanaka.demo.encodedcc.org/

forresttanaka committed:
ENCD-6191-cart-listing

forresttanaka commented:
I won’t be starting a QA demo now. I’ll start it Monday morning.

zoldello commented on src/encoded/static/components/cart/util.js:
```diff
@@ -205,3 +207,23 @@ export const getSeriesDatasets = (series, allDatasetsInCart) => (
+ *     released - cart is released by DCC
+ *     any - any of the above reasons apply
+ */
+export const getReadOnlyState = (cart) => {
```
Can this be simplified? All you are doing it creating three states - isLocked, isReleased, isCartStateSet. The name of the method had me thinking you were doing much more.

forresttanaka replied:
These are the potential reasons a cart might be locked, but there are unusual cases where a locked cart gets treated differently from a released cart, so they need distinguishing. For example, the Unlock button only gets disabled for a released cart, not a locked one.

zoldello commented on src/encoded/static/components/cart/add_multiple.js:
```diff
@@ -91,7 +93,7 @@ const CartAddAllSearchComponent = ({
                 <button
                     type="button"
-                    disabled={inProgress || savedCartObj.locked}
+                    disabled={inProgress || readOnlyState.any}
```
Why is this button disabled if the state is released (readOnlyState.any is truthy if release state is set to anything)?

forresttanaka replied:
The button gets disabled if the cart is released or locked, so “any.” readOnlyState.any is truth if the cart status is released or the cart is locked.

zoldello commented on src/encoded/static/components/cart/cart.js:
```diff
@@ -325,32 +331,19 @@ const requestDatasets = (elements, fetch, session) => {
     // button drops down a menu allowing the user to select the data type to view which links to
     // that report view.
-    if (linkedCart.current && linkedCart.current.elements && linkedCart.current.elements.length > 0) {
+    if (cart && cart.elements && cart.elements.length > 0) {
```
Can also be written as - 

if (cart?.elements?.length > 0) {

forresttanaka replied:
I noticed that just before code review; I had only converted existing code directly. I’ll change this if I decide to update the PR.

zoldello commented on src/encoded/static/components/cart/cart.js:
```diff
@@ -557,6 +541,8 @@ const CartSearchResultsControls = ({
 };
 
 CartSearchResultsControls.propTypes = {
+    /** Cart being displayed */
```
Is the comment correct?

forresttanaka replied:
No; more like, “The current cart.”

zoldello commented on src/encoded/static/components/cart/cart.js:
```diff
@@ -730,26 +716,30 @@ CounterTab.defaultProps = {
  * }
  */
 const getCartInfo = (context, savedCartObj) => {
+    let cart = null;
```
This is me nitpicking. When I see something set to null, it tells me you need the value to be null for use in the code. On the flip side, if the value is set later but no setting is needed at declaration, I have seen people not set it, which sets it to undefined by default (undefined tells me no value is needed at declaration).

forresttanaka replied:
I can see that argument. I’ll leave it since this is old code.

zoldello commented on src/encoded/static/components/cart/clear.js:
```diff
@@ -89,67 +89,61 @@ CartClearModal.contextTypes = {
-    handleClearCartClick() {
-        this.setState({ modalOpen: true });
-    }
+    const handleClearCartClick = () => {
```
In the event name, doesn't "Click" imply "handle"? Since you are in the cart file-folder, is it safe to assume the method supports a cart operation?

forresttanaka replied:
I keep changing how I name event handlers, but I think “ClearCart” helps here instead of just “Clear” which might still leave people wondering “Clear what? The cart? A field?”

zoldello replied:
Got point.

zoldello commented on src/encoded/static/components/cart/clear.js:
```diff
@@ -89,67 +89,61 @@ CartClearModal.contextTypes = {
+            <div className="cart-tools-extras__button">
+                <button
+                    type="button"
+                    disabled={inProgress || isCartReadOnly}
```
Do you need a disabled-setting for a system failure?

zoldello commented on src/encoded/static/components/cart/description.js:
```diff
@@ -0,0 +1,328 @@
+ * @param {string} description Description entered by the user
+ * @returns {string} `description` but sanitized and with preceding/trailing spaces removed
+ */
+export const sanitizeDescription = (description) => sanitize(description).trim();
```
Do you want to sanitize (as in remove DOM-like string) or ENCODE (prevent XSS attack by changing anything like DOM to a safe format)? This turns null to empty string, truthy to 'true', falsey to empty string. Are you okay with these?

forresttanaka replied:
It strips anything within a script tag, so it seems unlikely to ever come across anything truthy or falsy AFAIK.

zoldello commented on src/encoded/static/components/cart/description.js:
```diff
@@ -0,0 +1,328 @@
+ * @param {string} editedDescription Description as entered by the user before sanitizing
+ * @returns See return statement
+ */
+export const useCartDescription = (cart, editedDescription) => {
```
Is this a custom hook or just a regular function? Custom hooks I have seen typically wraps at least one of react inbuilt hooks.

forresttanaka replied:
Yeah, really just a regular function.

zoldello commented on src/encoded/static/components/cart/description.js:
```diff
@@ -0,0 +1,328 @@
+
+    React.useEffect(() => {
+        // Must have a DOM to sanitize the description.
+        setSanitizedDescription(marked(sanitizeDescription(description || '')));
```
sanitize takes care of falsey states for you.

zoldello commented on src/encoded/static/components/cart/description.js:
```diff
@@ -0,0 +1,328 @@
+    if (sanitizedDescription) {
+        return (
+            <div className="cart-description">
+                <div dangerouslySetInnerHTML={{ __html: sanitizedDescription }} />
```
If you took the trouble to remove security-risk markup, why now use dangerouslySetInnerHTML?

forresttanaka replied:
Is there another way?

forresttanaka replied:
I think it’s likely it was named `dangerouslySetInnerHTML` because you _could_ pass it executable code, not that it conjures executable code in markup that has been stripped of it.

zoldello replied:
It is called "dangerouslySetInnerHTML" because react designers believe an usual function name will force developer to put in some thought before using it. That is what I read about it. Now,  on my comment, my understanding is that sanitized function already took out the security-risk code. If that is true, do you need to use "dangerouslySetInnerHTML" to again sanitize the code. It also does not hurt to double secure security-risk code.

zoldello commented on src/encoded/static/components/cart/description.js:
```diff
@@ -0,0 +1,328 @@
+            <ModalBody addCss="cart-description-editor">
+                <p id="description-editor-description">
+                    This description gets saved with this cart, and appears along with this cart
+                    when you list it on the public <strong>Listed carts</strong> page. You can
```
<strong> is valid; but isn't it mixing markup with styling ..?

forresttanaka replied:
?

zoldello replied:
Wow. That came out wrong.  "strong tag" (\<strong\>) mixed html with css. Is that good practice?

zoldello commented on src/encoded/static/components/cart/description.js:
```diff
@@ -0,0 +1,328 @@
+                    when you list it on the public <strong>Listed carts</strong> page. You can
+                    format the text with Markdown syntax.
+                </p>
+                {cart.status === 'listed' &&
```
This is always true if the LHS is true... Right?

forresttanaka replied:
No real LHS here, as there’s no assignment. Do you mean the left of the `&&`? The `<p>` gets called if and only if `cart.status` has the value “listed.”

zoldello commented on src/encoded/static/components/cart/manager.js:
```diff
@@ -472,89 +463,62 @@ const NameCartButton = connect(NameCartButtonComponent.mapStateToProps, NameCart
-        this.handleCloseClick = this.handleCloseClick.bind(this);
-    }
+const DeleteCartButtonComponent = ({ cart, setInProgress, inProgress, fetch, updateCartManager }) => {
+    /** True if modal to confirm cart deletion open */
```
I think you mean - True if modal to confirm cart deletion is opened

zoldello commented on src/encoded/static/components/cart/menu.js:
```diff
@@ -0,0 +1,215 @@
+ * Renders the cart icon menu and count or in-progress spinner in the nav bar.
+ */
+const CartNavTitle = ({ elements, inProgress, locked }) => {
+    let status;
```
Does status === undefined mean no status is set?

forresttanaka replied:
I originally called the little icon a cart “status” which now is confusing since the cart’s `status` property has new meaning. `status === undefined` means no little icon to display.

zoldello commented on src/encoded/static/components/cart/menu.js:
```diff
@@ -0,0 +1,215 @@
+
+    // Define the menu items for the Cart Status menu.
+    const viewCartItem = <a key="view" href="/cart-view/">View cart</a>;
+    const clearCartItem = !savedCartObj.locked && <button type="button" key="clear" onClick={clearCartClick}>Clear cart</button>;
```
Would **nullish coalescing** be a good option in lieu of this pattern?

forresttanaka replied:
Not sure how that would help. If the cart isn’t locked, then it displays the “Clear cart” button.

zoldello commented on src/encoded/static/components/cart/menu.js:
```diff
@@ -0,0 +1,215 @@
+    const menuItems = [];
+    if (loggedIn) {
+        // Build the disabled cart name and lock status.
+        const cartName = (savedCartObj && savedCartObj.name) && truncateString(savedCartObj.name, 22);
```
Why magic number 22?

zoldello commented on src/encoded/static/components/cart/search_listing.js:
```diff
@@ -0,0 +1,141 @@
+
+/**
+ * Display the description of a cart in the cart search results. Include a control to collapse and
+ * expand the description when the description has more text than would fit in the allowed vertial
```
"vertical". It would be cool if we had the power like Shakespeare to invent new words rather than limiting ourselves to a dictionary.

zoldello commented on src/encoded/static/components/cart/share.js:
```diff
@@ -1,20 +1,30 @@
  * Components to display the Share Cart modal.
  */
+
+// node_modules
```
These should be obvious or am I mistaken...?

forresttanaka replied:
I’ve started to do this to make it more obvious that some imports are out of order, not that being out of order is a technical issue, but I just prefer not to have library imports below local ones, yet I was doing that sometimes. These comments help me prevent that.

zoldello replied:
Let's talk about this off-line sometimes. Imports cannot be out of order as package management handles that for you. Developers should get an error if they try to do something like call an internal package before an external one.

zoldello commented on src/encoded/static/components/cart/store.js:
```diff
@@ -0,0 +1,215 @@
+ */
+const generateFileViewsMinusFiles = (fileViews, title, filePaths) => {
+    // Find file view with the given title.
+    const indexOfFileViewWithMatchingTitle = fileViews.findIndex((view) => view.title === title);
```
Maybe - fileViewIndexWithMatchingTitle

zoldello commented on src/encoded/static/components/objectutils.js:
```diff
@@ -1178,3 +1178,26 @@ export const collectDatasetBiosamples = (dataset) => {
+ */
+export const TextAreaCounter = ({ value, maxLength, css }) => (
+    <div className={`textarea-counter${css && ` ${css}`}`}>
+        {value.length} / {maxLength}
```
Do you need to Math. {floor/ceil} this so you always get a whole number?

forresttanaka replied:
This is a literal slash; not division, e.g. “50 / 1000.”

zoldello replied:
Opps. Thanks.

zoldello commented on src/encoded/static/scss/encoded/modules/_cart.scss:
```diff
@@ -1660,9 +1671,229 @@ $file-view-checkbox-size: 14px;
+    }
+
+    ul, ol {
+        margin: 0.3em;
```
should we be using "em"?

forresttanaka replied:
I copied this from a post with a URL I forgot to include that was for resetting list styles. So, I’m not sure in this case; I presume this had been tested, and I tested it as well. Not sure if there are cases this wouldn’t work.

zoldello reviewed (dismissed):
None of my comment are must-have. A few are nitpicking.

forresttanaka committed:
ENCD-6191-cart-sharing

Address code-review feedback.

forresttanaka committed:
ENCD-6191-cart-sharing

Fix ESLint error.

zoldello reviewed (dismissed):
Code looks fine. My comments are philosophical and don't matter. The last two changes are cosmetic; so nothing significant changed.

forresttanaka committed:
ENCD-6191-cart-sharing

Correct cart status display to use the new `cart` variable that takes shared and active carts into account, instead of the old `savedCartObj` that only displays the active cart’s status.

forresttanaka committed:
ENCD-6191-cart-sharing

Change old “Listing” button to say “List cart” or “Unlist cart” as appropriate.

forresttanaka commented:
The cart status used the current cart object instead of the normalized `cart` object so it could display the wrong thing on shared-cart displays. It now always shows the normalized `cart` object status. I also changed the “Listing” button to say “List cart” for unlisted carts, or “Unlist cart” for listed carts.

forresttanaka commented:
Preview demo with the two new changes: https://encd-6191-cart-sharing-pv2-fytanaka.demo.encodedcc.org/

forresttanaka committed:
ENCD-6191-cart-sharing

Update BDD test for carts to adjust to new “List cart” button name.

zoldello commented on src/encoded/tests/features/cart.feature:
```diff
@@ -82,7 +82,7 @@ Feature: Cart
         Then I should not see an element with id "description-editor"
 
     Scenario: Cart listing
-        When I press "Listing"
```
Do you need to test for unlisted?

zoldello reviewed (approved):
Code looks good. My comment is not significant.
